### ISSUE TITLE
rc: need a way for 3rd party packages to override their own rc scripts

### ISSUE THREAD

**Issue by grondo** [flux-framework/flux-core#2788]
It was found in  flux-framework/flux-sched#607 that a flaw in rc init design is that flux-framework subprojects that install rc scripts will end up running two copies of their rc scripts under `make check` when they their is an earlier version of the package already installed to the same prefix.

This is because there is no way for the subpackage to override the older, installed rc scripts in `rc1.d`,`rc3.d` with newer versions being tested, and added via `FLUX_RC_PATH`.

There are a couple suggested solutions from flux-framework/flux-sched#607:

 * implement a rc naming scheme so that flux-core can associate rc scripts with a package, and an environment variable to "skip" all *installed* versions of scripts that match the name.
 * implement a facility to guard against running the same script multiple times. This might require sourcing rc scripts instead of running them, and change the order such that `FLUX_RC_EXTRA` scripts run before the "core" `rc1.d` scripts.

**Comment by grondo** [comment]
A potential solution for the 2nd bullet:

```diff
diff --git a/etc/rc1 b/etc/rc1
index 800f15bd2..70984939b 100755
--- a/etc/rc1
+++ b/etc/rc1
@@ -39,14 +39,26 @@ flux module load sched-simple & pids+=($!)
 wait_check ${pids[@]}
 unset pids
 
+declare -A FLUX_RC_INCLUDED
+flux_rc_included()
+{
+    local name=$1
+    if test ${FLUX_RC_INCLUDED[$name]}; then
+        echo "rc $name already included skipping ${BASH_SOURCE[1]}"
+        return 0
+    fi
+    FLUX_RC_INCLUDED[$name]=t
+    return 1
+}
+
 core_dir=$(cd ${0%/*} && pwd -P)
-all_dirs=$core_dir${FLUX_RC_EXTRA:+":$FLUX_RC_EXTRA"}
+all_dirs=${FLUX_RC_EXTRA:+"$FLUX_RC_EXTRA:"}$core_dir
 IFS=:
 shopt -s nullglob
 for rcdir in $all_dirs; do
     for rcfile in $rcdir/rc1.d/*; do
-       echo running $rcfile
-        $rcfile
+        echo sourcing $rcfile
+        source $rcfile
     done
 done
 shopt -u nullglob
```

Then in any rc script which you want to guard you could add
```shell
flux_rc_included myname && return 0
```

To the top of the script. E.g.

```console
grondo@asp:~/git/f.git$ cat etc/rc1.d/01-test 
flux_rc_included test && return 0

echo "In test script"
grondo@asp:~/git/f.git$ cat etc/rc1.d/02-test 
flux_rc_included test && return 0

echo "In test script 2"
```

```
2020-03-01T17:37:29.286504Z broker.info[0]: rc1: sourcing /home/grondo/git/f.git/etc/rc1.d/01-enclosing-instance
2020-03-01T17:37:29.302965Z broker.info[0]: rc1: sourcing /home/grondo/git/f.git/etc/rc1.d/01-test
2020-03-01T17:37:29.303081Z broker.info[0]: rc1: In test script
2020-03-01T17:37:29.303110Z broker.info[0]: rc1: sourcing /home/grondo/git/f.git/etc/rc1.d/02-test
2020-03-01T17:37:29.303158Z broker.info[0]: rc1: rc test already included skipping /home/grondo/git/f.git/etc/rc1.d/02-test
```

However, there are some drawbacks to this approach

 * switch to `source` for rc scripts could be perilous
 * switch to `source` requires all rc scripts be bash scripts
 * not sure about the change in order of FLUX_RC_EXTRA

**Comment by dongahn** [comment]
Of all the drawbacks, maybe we need to discuss this a bit more.

> not sure about the change in order of FLUX_RC_EXTRA

This will change the behavior between installed runs vs. in-tree runs. If nothing else, when FLUX_RC_EXTRA is added, `qmanager` will run first and then `sched-simple` run next and fail. 

Maybe something like this isn't that bad because their party packages can override some core's services. But as is, the core's RC also needs to be cognizant of overriding 3rd party package so probably not good?

**Comment by grondo** [comment]
> This will change the behavior between installed runs vs. in-tree runs. If nothing else, when FLUX_RC_EXTRA is added, qmanager will run first and then sched-simple run next and fail.

Not currently, since `sched-simple` is loaded from the main `rc1` script and not from `rc1.d/*`. The only thing in `rc1.d` currently is `01-enclosing-instance`.

**Comment by dongahn** [comment]
> Not currently, since sched-simple is loaded from the main rc1 script and not from rc1.d/*. The only thing in rc1.d currently is 01-enclosing-instance.

Ah. I didn't catch that. Sorry. Then, the side effect shouldn't be that bad at all...

**Comment by grondo** [comment]
It does indeed introduce a difference in ordering between normal startup vs startup under a testsuite that introduces FLUX_RC_EXTRA. That is something to think about. The naming scheme you had proposed would be simpler (though actually perhaps similar ordering changes would result,  since the extra scripts would be "skipped" in their normal order and run after everything else...)

**Comment by dongahn** [comment]
Yeah... Maybe order change cannot be avoided easily because we are "overriding."

The other thing we can do is to do the strongest overriding...That is, the installed third party `RCs` run and is loaded first. But then in the RC scripts themselves, we can add `flux remove -f <module>` first so that they can "replace".

This comes at a performance cost, but still doesn't introduce any of the disadvantage you identified... And no changes to the core.

**Comment by grondo** [comment]
> we can add flux remove -f <module> first so that they can "replace".

You could also do `flux module reload -f <module>` to load a module, removing any existing module of the same name first.

**Comment by dongahn** [comment]
That sounds even better.

**Comment by grondo** [comment]
This will work if all the 3rd party rc scripts do is load/unload modules. However, if the rc scripts do other generic work, like setup something in kvs, set/change broker attributes, etc., then we'll still need a way to override the installed scripts when we want to run the 3rd party framework project "out of the build directory".

However, with the stopgap solution above, perhaps resolution of this issue isn't quite as urgent.

**Comment by dongahn** [comment]
Agreed with both of your points.

 I think it would be a good goal a third party modules strive to make their clean load and unload like this state idempotent.
